### ISSUE TITLE
Restoring the model from the snapshot file.

### ISSUE THREAD

**Issue by nd7141** [catboost/catboost#479]
Problem: Is it possible to restore the model for _prediction_ from the training snapshot? As far as I understand it's possible to recover the model for the _training_, if [the file exists](https://tech.yandex.com/catboost/doc/dg/features/snapshots-docpage/#snapshots). But in case of `predict` method, there is no argument `shapshot_file`, and I think the only way is to use `load_model` method (which I think is possible only if the training is finished successfully). I tried to provide a snapshot file to the `fit` method with 0 iterations, but it fails saying the number of iterations should be positive. 
catboost version: 0.10.2
Operating System: Ubuntu
CPU: CPU

**Comment by annaveronika** [comment]
To do that you need to run training with the same parameters using this snapshot. Then the training will just use the snapshot and build a model from that. After that you can make predictions.

**Comment by nd7141** [comment]
Sorry, maybe I don't understand it correctly. I have the following use case, I have the time limit on training, say , 1 minute, I want to train the model for 1 minute, and then do predictions. I thought it's possible to do so by training for many iterations and then when the model stops, because of the time limit, use the latest snapshot to do predictions. Is it possible? 

I'm aware of `save/loaf` methods, but they are possible only when the model successfully finished the training.

**Comment by annaveronika** [comment]
After your training stops by timeout, you can watch in log, how many iterations have passed and train the model with this amount of iterations on top of the snapshot.
Btw snapshots are now saved by default every 5 minutes, I think, so you need to set the interval to something less.

**Comment by annaveronika** [comment]
@nd7141 did this help?

**Comment by nd7141** [comment]
Sorry, I'm being unclear. Imagine the following use case. Your data arrives every 5 minutes, and rewrites the previous one. During these 5 minutes you want to train the model as much as you can. Ideally, I would like to set a parameter `timeout=5` and then use the model after that. I didn't find such parameter (but maybe there exists one?). Another option I see is to set the number of iterations high enough, say 1M, and then when the model training fails (because of new data) I would use the latest snapshot. However, if I restore the latest snapshot, I need to train the model again, which I don't want (because I don't have training data anymore) -- I just want to use the trained model to make predictions. So the question is how can I train the model to the maximum and then use it for making predictions?

**Comment by annaveronika** [comment]
The solution to your problem is the following: you train the model and write stdout to file. After the training has stopped by timeout, you take a look into the log, see how many iterations have passed and run training using this number of iterations.

**Comment by nd7141** [comment]
Thank you @annaveronika I agree that could be a solution when the period of update is constant and data amount does not change. But with every new training session, you get new data which updated not exactly at the same period. Maybe it's possible to add `timeout` parameter to `fit` method?

**Comment by annaveronika** [comment]
In your case you need to write a script that reads the logs and parses amount of iterations from it. The simplest thing would be to look on count of lines in file with test errors or train errors.
This script will parse count of lines and then run training using this value.

I don't think it's a good idea to add timeout to fit method, because it is a very specific case when you need it, I don't see other scenarios when it will be useful. And in your case there is a different solution, so you can try to use that one.

**Comment by nd7141** [comment]
Well, I think it's common industry setup: you get data at some point of time (you monitor for it), and then you train your model until new data arrive for as much as you can. You can think of other scenarios when you send your code to evaluator, who then executes it under his own time and resource constraints, to which you don't have access. Making it easy to control the amount of time your model is trained can boost the performance in these cases. 
Having said that, your solution to the issue, while cumbersome, works, so it's your call whether to add it or not.

**Comment by annaveronika** [comment]
Usually it makes sense to train the model until overfitting instead (you can use larger learning rate + overfitting detector for that). Otherwise you will get an underfitted model which will not give you good results.

**Comment by annaveronika** [comment]
Ok, I'm happy the solution did work for you! Let us know if you have any other problems.
